okhttp/third_party/okhttp/java/io/grpc/okhttp/internal/Platform.java

```diff
@@ -105,47 +119,71 @@ public void connectSocket(Socket socket, InetSocketAddress address,
   }
 
   /** Attempt to match the host runtime to a capable Platform implementation. */
+  @SuppressWarnings("unchecked")
   private static Platform findPlatform() {
-    // Attempt to find Android 2.3+ APIs.
+    // Find the conscrypt security provider if we can
+    Class<Provider> conscryptProviderClass = null;
     try {
+      conscryptProviderClass =
+          (Class<Provider>) Class.forName("org.conscrypt.OpenSSLProvider");
+    } catch (Throwable t) {
       try {
-        Class.forName("com.android.org.conscrypt.OpenSSLSocketImpl");
-      } catch (ClassNotFoundException e) {
-        // Older platform before being unbundled.
-        Class.forName("org.apache.harmony.xnet.provider.jsse.OpenSSLSocketImpl");
+        conscryptProviderClass =
+            (Class<Provider>)  Class.forName("com.android.org.conscrypt.OpenSSLProvider");
+      } catch (Throwable t2) {
+        try {
+          conscryptProviderClass = (Class<Provider>)
+              Class.forName("org.apache.harmony.xnet.provider.jsse.OpenSSLProvider");
+        } catch (Throwable t3) {
+          // Stick with what we have
+        }
+      }
+    }
+    boolean haveConscrypt = false;
+    Provider sslProvider = null;
+    try {
+      sslProvider = SSLContext.getDefault().getProvider();
+    } catch (NoSuchAlgorithmException nsae) {
+      // Ignore
+    }
+    if (conscryptProviderClass != null) {
+      if (sslProvider != null && sslProvider.getClass().equals(conscryptProviderClass)) {
+        haveConscrypt = true;
+      } else {
+        try {
+          sslProvider = conscryptProviderClass.newInstance();
+          haveConscrypt = true;
+        } catch (Throwable t) {
+          // Unable to use conscrypt, fall through to Jetty
+          logger.log(Level.WARNING,
+              "Unable to create conscrypt provider " + conscryptProviderClass.getName(), t);
+        }
       }
+    }
 
+    if (haveConscrypt) {
+      // Attempt to find Android 2.3+ APIs.
       OptionalMethod<Socket> setUseSessionTickets
           = new OptionalMethod<Socket>(null, "setUseSessionTickets", boolean.class);
       OptionalMethod<Socket> setHostname
           = new OptionalMethod<Socket>(null, "setHostname", String.class);
       Method trafficStatsTagSocket = null;
```

Comment:
Are these still useful now? Seems all we care about is getAlpnSelectedProtocol and setAlpnSelectedProtocol. And after all the checks for conscrypt, we don't even care about the Android version.